### ISSUE TITLE
Make appropriate use of ToList and ToArray

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFramework#702]
TL;DR Use ToArray for long-lived IReadOnlyList; Use ToList when short-lived

Relevant conversation with David:

[avickers]: Curious: Is ToList generally faster than using ToArray? How about using an array verses a List after it is constructed? 
[daobando]: The performance of an array and a list after they are constructed are basically the same (a few more assembly instructions for the list, but not enough to cause noticeable difference). Internally, a list has one or more arrays holding all the elements and a few open spots for new insertions; it builds a new larger array once a capacity threshold is reached but in terms of access performance it's basically the same as an array.
In this particular case, the call to IEnumerable.ToArray() or IEnumerable.ToList() is meant to generate a short-lived object. When one calls ToArray(), an arbitrary size array is generated simply because IEnumerable doesn't have a Count property. The array may be re-generated a number of times to accommodate for extra elements (same as a List would do) but at the very end, a final allocation is done to return an array with only the exact element count we have.
If it were a long-lived element, I'd prefer calling ToArray() since it gives the best use of memory long term. But since this is just a temporary variable, and we care about performance in this code path, reducing the amount of allocations is preferable to saving memory. 
[avickers]: Cool. We should probably use ToList more often in the new stack where I have been using ToArray quite a lot since some perf benchmarking I did a number of years ago indicated it was faster, but this may not have been the short-lived scenario.
